### PR TITLE
Decode raw records if possible in python stream reader

### PR DESCRIPTION
**Pull request by foxymiles**
**Public-Facing Changes**
This improves the ergonomics of the stream reader class.

**Description**
This adds a `decoded_records` method to the stream reader class that returns only decoded messages instead of all the lower level mcap records if the required ros1 libraries are available. We can expand on this later to support encodings other than ros1.

### CODE REVIEW

**Comment by jtbandes** [comment]
I don't think we should structure our Python MCAP lib with an (even optional) dependency on `genpy` or other ROS stuff. Could we make it a separate package instead? Like change the iterator to accept a "decoder" function that the user can configure?

**Comment by foxymiles** [comment]
> I don't think we should structure our Python MCAP lib with an (even optional) dependency on `genpy` or other ROS stuff. Could we make it a separate package instead? Like change the iterator to accept a "decoder" function that the user can configure?

The goal here is make this as easy to use for casual users as possible. I guess if we also supplied the ros1 decoder that might be an ok solution.

**Comment by jtbandes** [comment]
Sometimes explicit composition is easier in the long run than things that work "magically" :)

**Comment by jhurliman** [comment]
I don't understand the terminology here. What is a "raw record"? ROS doesn't use the term record, only MCAP, and we decode all the MCAP record types already. The PR description says:

> This adds a decoded_records method to the stream reader class that returns only decoded messages

It sounds like messages (or message payloads) and records are being used as interchangeable terms here.

File: python/mcap/mcap0/stream_reader.py [hunk 1]
```diff
@@ -96,6 +102,37 @@ def records(self) -> Iterator[McapRecord]:
...
+                    msg_type = dynamic.generate_dynamic(  # type: ignore
+                        record.name, record.data.decode()
+                    )
+                    msg_types[record.name] = msg_type[record.name]
+            if isinstance(record, Channel):
+                channels[record.id] = record
+            if isinstance(record, Message):
+                channel = channels[record.channel_id]
+                schema = schemas[channel.schema_id]
+                message = msg_types[schema.name]().deserialize(record.data)
```
Comment (jhurliman) [inline python/mcap/mcap0/stream_reader.py:133]: What happens if this fails?

**Comment by foxymiles** [comment]
> I don't understand the terminology here. What is a "raw record"? ROS doesn't use the term record, only MCAP, and we decode all the MCAP record types already. The PR description says:
> 
> > This adds a decoded_records method to the stream reader class that returns only decoded messages
> 
> It sounds like messages (or message payloads) and records are being used as interchangeable terms here.

Yeah something like `decoded_messages` would be a better name here.

File: python/mcap/mcap0/stream_reader.py (hunk 1 above)
Comment (foxymiles) [reply python/mcap/mcap0/stream_reader.py:133]: Should probably wrap this in a try/catch that throws an `McapError` right?

File: python/mcap/mcap0/writer.py [hunk 2]
```diff
@@ -48,9 +48,9 @@ class Writer:
 
     def __init__(
         self,
-        output: Union[str, BytesIO, BufferedWriter],
+        output: Union[str, IO[Any]],
         chunk_size: int = 1024 * 1024,
-        compression: CompressionType = CompressionType.NONE,
+        compression: CompressionType = CompressionType.ZSTD,
```
Comment (wkalt) [inline python/mcap/mcap0/writer.py:53]: I think this change should definitely be made, maybe independent of the larger patch

Comment (foxymiles) [reply python/mcap/mcap0/writer.py:53]: Yeah I made this in separate smaller PR. I'm going to close this PR in favor of another package that provides a separate ROS1 decoder class.

**Comment by foxymiles** [comment]
Closed in favor of https://github.com/foxglove/python-mcap-ros1-support

**Pull request closed without merging**